### PR TITLE
core: Add PrivateUsers=full

### PR DESCRIPTION
**Pull request by ryantimwilson**
Recently, PrivateUsers=identity was added to support mapping the first 65536 UIDs/GIDs from parent to the child namespace and mapping the other UID/GIDs to the nobody user.

However, there are use cases where users have UIDs/GIDs > 65536 and need to do a similar identity mapping. Moreover, in some of those cases, users want a full identity mapping from 0 -> UID_MAX.

To support this, we add PrivateUsers=full that does identity mapping for all available UID/GIDs.

Note to differentiate ourselves from the init user namespace, we need to set up the uid_map/gid_map like:
```
0 0 1
1 1 UINT32_MAX - 1
```

as the init user namedspace uses `0 0 UINT32_MAX` and some applications - like systemd itself - determine if its a non-init user namespace based on uid_map/gid_map files.

Note systemd will remove this heuristic in running_in_userns() in version 258 (https://github.com/systemd/systemd/pull/35382) and uses namespace inode. But some users may be running a container image with older systemd < 258 so we keep this hack until version 259 for version N-1 compatibility.

In addition to mapping the whole UID/GID space, we also set /proc/pid/setgroups to "allow". While we usually set "deny" to avoid security issues with dropping supplementary groups (https://lwn.net/Articles/626665/), this ends up breaking dbus-broker when running /sbin/init in full OS containers.

Fixes: #35168
Fixes: #35425

### CODE REVIEW

**Review by DaanDeMeyer — APPROVED** [review]

File: src/core/exec-invoke.c [hunk 1]
```diff
@@ -2103,6 +2103,22 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
...
+                 * comparing uid_map/gid_map to `0 0 UINT32_MAX`. Thus, we still map all UIDs/GIDs but do it
+                 * using two extents to differentiate the new user namespace from the init namespace:
+                 *   0 0 1
+                 *   1 1 UINT32_MAX - 1
+                 *
+                 * Note the kernel defines the UID range between 0 and UINT32_MAX so we map all UIDs even though
+                 * the UID range beyond INT32_MAX (e.g. i.e. the range above the signed 32-bit range) is
+                 * icky. For example, setfsuid() returns the old UID as signed integer. But units can decide to
+                 * use these UIDs/GIDs so we need to map them. */
+                r = asprintf(&uid_map, "0 0 1\n1 1 " UID_FMT "\n", UINT32_MAX - 1);
```
Comment (poettering) [inline src/core/exec-invoke.c]: please format this with a line break, i.e:

```c
r = asprintf(&uid_map, "0 0 1\n"
                       "1 1 " UID_FMT "\n", UINT32_MAX - 1);
```

File: src/core/exec-invoke.c [hunk 2]
```diff
@@ -2123,6 +2139,10 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
                 gid_map = strdup("0 0 65536\n");
                 if (!gid_map)
                         return -ENOMEM;
+        } else if (private_users == PRIVATE_USERS_FULL) {
+                r = asprintf(&gid_map, "0 0 1\n1 1 " UID_FMT "\n", UINT32_MAX - 1);
```
Comment (poettering) [inline src/core/exec-invoke.c]: ditto

**Comment by bluca** [comment]
This is a new option so it's really material for 258

**Comment by ryantimwilson** [comment]
@bluca yea waiting for 258 is fine. I assumed there was another 257 branch but guess we just use main, then branch after all RC releases? Anyways, I'm fine to wait

**Comment by ryantimwilson** [comment]
@poettering @DaanDeMeyer I'm requesting a re-review from you since this PR isn't merged yet.

I added a second commit in the PR to fix: https://github.com/systemd/systemd/issues/35425. TL;DR: PrivateUsers=full also sets /proc/pid/setgroups = "allow"

I decided against adding another setting ProtectGetGroups since 1) It would be duplicative of SystemCallFilter (in terms of result) 2) The semantics of ProtectGetGroups= were confusing since it was so highly tied to user namespaces

Happy to hear other ideas/approaches!

File: src/core/exec-invoke.c [hunk 3]
```diff
@@ -4979,7 +5007,9 @@ int exec_invoke(
                 if (pu == PRIVATE_USERS_NO)
                         pu = PRIVATE_USERS_SELF;
 
-                r = setup_private_users(pu, saved_uid, saved_gid, uid, gid);
+                /* The kernel requires /proc/pid/setgroups be set to "deny" prior to writing /proc/pid/gid_map in
+                 * unprivileged user namespaces. */
+                r = setup_private_users(pu, saved_uid, saved_gid, uid, gid, false /*allow_setgroups=*/);
```
Comment (yuwata) [inline src/core/exec-invoke.c]: Spurious order of the short comment: `/*allow_setgroups=*/ false`

File: src/core/exec-invoke.c [hunk 4]
```diff
@@ -5171,7 +5179,8 @@ int exec_invoke(
          * different user namespace). */
 
         if (needs_sandboxing && !userns_set_up) {
-                r = setup_private_users(context->private_users, saved_uid, saved_gid, uid, gid);
+                r = setup_private_users(context->private_users, saved_uid, saved_gid, uid, gid,
+                                        context->private_users == PRIVATE_USERS_FULL);
```
Comment (yuwata) [inline src/core/exec-invoke.c]: Please also add short comment for the boolean: `/* allow_setgroups = */`

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by ryantimwilson** [comment]
@yuwata nice, didn't see that PR, yes lets remove the hack then

The only gotcha is while Meta deploys 258, we will be booting containers with 257 (aka without https://github.com/systemd/systemd/pull/35382). But we can always add a hotpatch on our Meta build to use the old hack - agree that upstream doesn't need to support the hack

**Comment by ryantimwilson** [comment]
@DaanDeMeyer @yuwata maybe a compromise - we can keep the old version but add a TODO to remove this in 259?

**Comment by yuwata** [comment]
> > @yuwata nice, didn't see that PR, yes lets remove the hack then
> > The only gotcha is while Meta deploys 258, we will be booting containers with 257 (aka without #35382). But we can always add a hotpatch on our Meta build to use the old hack until we're fully on 258 - agree that upstream doesn't need to support the hack
> 
> Frankly I find this a good argument to keep the previous approach. The code is trivial and it makes sure we keep compat with older versions of systemd running in containers or other sandboxes that still use the old heuristics. @yuwata What do you think?

Not a strong opinion, but I do not think it is necessary. The PR #35382 just drops old heuristic logic about userns. In v257, we already have a2429f507ca72278bb6d2893b58906ab313d6e0c, and that should enough to detect non-init userns.
If you want to only run v257 (or newer) containers on `PrivateUsers=full`, then such workaround is not necessary. The workaround is only necessary If you want to run not-patched v256 (or older) containers.

**Comment by ryantimwilson** [comment]
@yuwata the problem with https://github.com/systemd/systemd/commit/a2429f507ca72278bb6d2893b58906ab313d6e0c#diff-d5b6c9503f12e3cfcc77bbdc2e1184b772ccedbaa2826605a5a3dae0dbd9149fR893 in a container with PrivateUsers=full is that `namespace_is_init(NAMESPACE_USER)` will return `0` and we will fall through to the other heuristics and will do the uid_map/gid_map check. https://github.com/systemd/systemd/pull/35382 will fix this correctly by removing the other heuristics.

So unfortunately, this will affect containers with 256+257.

If you don't mind, I'll revert to the previous hack but add a comment on why and to remove this in 259.

**Comment by yuwata** [comment]
Ah, indeed. Bad... OK. Feel free to keep the maps as previous.

File: src/core/exec-invoke.c [hunk 5]
```diff
@@ -2103,6 +2103,29 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
...
+                 * this uid_map/gid_map hack until version 259 for version N-1 compatibility.
+                 *
+                 * TODO: Switch to `0 0 UINT32_MAX` in systemd v259.
+                 *
+                 * Note the kernel defines the UID range between 0 and UINT32_MAX so we map all UIDs even though
+                 * the UID range beyond INT32_MAX (e.g. i.e. the range above the signed 32-bit range) is
+                 * icky. For example, setfsuid() returns the old UID as signed integer. But units can decide to
+                 * use these UIDs/GIDs so we need to map them. */
+                r = asprintf(&uid_map, "0 0 1\n"
+                                       "1 1 " UID_FMT "\n", UINT32_MAX - 1);
```
Comment (YHNdnzj) [inline src/core/exec-invoke.c]: Please cast `UINT32_MAX - 1` to uid_t. We prefer being explicit for typedef-ed types.

File: src/core/exec-invoke.c [hunk 6]
```diff
@@ -2123,6 +2146,11 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
                 gid_map = strdup("0 0 65536\n");
                 if (!gid_map)
                         return -ENOMEM;
+        } else if (private_users == PRIVATE_USERS_FULL) {
+                r = asprintf(&gid_map, "0 0 1\n"
+                                       "1 1 " UID_FMT "\n", UINT32_MAX - 1);
```
Comment (YHNdnzj) [inline src/core/exec-invoke.c]: GID_FMT and cast to gid_t

File: src/core/exec-invoke.c [hunk 7]
```diff
@@ -5149,7 +5185,8 @@ int exec_invoke(
          * different user namespace). */
 
         if (needs_sandboxing && !userns_set_up) {
-                r = setup_private_users(context->private_users, saved_uid, saved_gid, uid, gid);
+                r = setup_private_users(context->private_users, saved_uid, saved_gid, uid, gid,
+                                        /*allow_setgroups=*/ context->private_users == PRIVATE_USERS_FULL);
```
Comment (YHNdnzj) [inline src/core/exec-invoke.c]:
```suggestion
                                        /* allow_setgroups= */ context->private_users == PRIVATE_USERS_FULL);
```

**Review by YHNdnzj — APPROVED** [review]

**Comment by DaanDeMeyer** [comment]
I made the logic of the second commit slightly cleaner (It's what I would have requested in a review if ryan was available, but as he isn't I just pushed the fix to this PR). Setting the green label.

**Review by bluca — APPROVED** [review]

**Pull request merged**